File: 백준/Gold/9663.N－Queen/N－Queen.py
def check(r):
    for i in range(r):
        if q[i] == q[r] or abs(q[r] - q[i]) == r - i:
            return False
    return True

def dfs(r):
    global res
    if r == n:
        res += 1
        return
    for i in range(n):
        q[r] = i
        if check(r):
            dfs(r + 1)

n = int(input())
q = [0] * n
res = 0
dfs(0)
print(res)

File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <bits/stdc++.h>
using namespace std;

int n;
int res = 0;
int q[15] = {0,};

bool check(int r)
{
  for (int i = 0; i < r; i++)
    if (q[i] == q[r] || abs(q[r] - q[i]) == r - i )
      return false;
  return true;
}

void qeens(int r)
{
  if (r == n)
  {
    res++;
    return;
  }

  for (int i = 0; i < n; i++)
  {
    q[r] = i;
    if (check(r))
      qeens(r + 1);
  }
}

int main()
{
  cin >> n;
  qeens(0);
  cout << res;
}

